### PR TITLE
UAR-611: Handling missing OE, MO and BO data

### PR DESCRIPTION
**Pull request by bbell-companieshouse**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-611

### Change description
There could be situations where not all public and/or private data can be found for OE/BO/MO

### Work checklist

- [ ] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java [hunk 1]
```diff
@@ -75,12 +73,12 @@ public List<Change> collateBeneficialOwnerChanges(
...
    * list is returned.
    */
-  private List<Change> individualBeneficialOwnerChange() {
+  private List<Change> individualBeneficialOwnerChange(Map<String, Object> logMap) {
     var beneficialOwnersIndividual = overseasEntitySubmissionDto.getBeneficialOwnersIndividual();
     return beneficialOwnersIndividual
         .stream()
         .filter(bo -> bo.getChipsReference() != null)
-        .map(this::covertBeneficialOwnerIndividualToChange)
+        .map(individualBO -> covertBeneficialOwnerIndividualToChange(individualBO, logMap))
```
Comment (bmcgreevyCH) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java:81]: Would you mind doing a search on the whole file for the word COVERT and change them to CONVERT? Cheers.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerCessationService.java [hunk 2]
```diff
@@ -155,6 +164,16 @@ private String getAppointmentId(Pair<PscApi, PrivateBoDataApi> publicPrivateBoPa
 
   private Optional<Pair<PscApi, PrivateBoDataApi>> getPublicPrivateBoPairOptional(
           String hashedId, Map<String, Pair<PscApi, PrivateBoDataApi>> combinedBoData) {
-    return Optional.ofNullable(combinedBoData.get(hashedId));
+    return Optional.ofNullable(combinedBoData).map(data -> data.get(hashedId));
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerCessationService.java:167]: Good fix

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java [hunk 3]
```diff
@@ -150,10 +149,20 @@ private BeneficialOwnerChange<CorporateBeneficialOwnerPsc> covertBeneficialOwner
...
 
-    if (publicPrivateBoPair == null || publicPrivateBoPair.getRight() == null) {
-      ApiLogger.errorContext(SERVICE, NO_PSC_FOUND, null);
+    if (publicPrivateBoPair == null) {
+      ApiLogger.errorContext(SERVICE, NO_PUBLIC_AND_NO_PRIVATE_BO_DATA_FOUND, null, logMap);
       return null;
     }
+    else {
+      if (publicPrivateBoPair.getLeft() == null) {
+        ApiLogger.errorContext(SERVICE, NO_PUBLIC_BO_DATA_FOUND, null, logMap);
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java:158]: Maybe this case (and similar) below could say "continuing" somehow? As we don't return null;

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java [hunk 4]
```diff
@@ -224,18 +233,30 @@ private BeneficialOwnerChange<CorporateBeneficialOwnerPsc> covertBeneficialOwner
...
-    if (publicPrivateBoPair == null || publicPrivateBoPair.getRight() == null) {
-      ApiLogger.errorContext(SERVICE, NO_PSC_FOUND, null);
+
+    if (publicPrivateBoPair == null) {
+      ApiLogger.errorContext(SERVICE, NO_PUBLIC_AND_NO_PRIVATE_BO_DATA_FOUND, null, logMap);
       return null;
     }
+    else {
+      if (publicPrivateBoPair.getLeft() == null) {
+        ApiLogger.errorContext(SERVICE, NO_PUBLIC_BO_DATA_FOUND, null, logMap);
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java:252]: similar to above

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java [hunk 5]
```diff
@@ -303,10 +325,20 @@ private BeneficialOwnerChange<IndividualBeneficialOwnerPsc> covertBeneficialOwne
...
 
-    if (publicPrivateBoPair == null || publicPrivateBoPair.getRight() == null) {
-      ApiLogger.errorContext(SERVICE, NO_PSC_FOUND, null);
+    if (publicPrivateBoPair == null) {
+      ApiLogger.errorContext(SERVICE, NO_PUBLIC_AND_NO_PRIVATE_BO_DATA_FOUND, null, logMap);
       return null;
     }
+    else {
+      if (publicPrivateBoPair.getLeft() == null) {
+        ApiLogger.errorContext(SERVICE, NO_PUBLIC_BO_DATA_FOUND, null, logMap);
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java:334]: same / say continuing?

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/ManagingOfficerChangeService.java [hunk 6]
```diff
@@ -198,18 +209,29 @@ private ManagingOfficerChange<IndividualManagingOfficer> convertManagingOfficerI
...
 
-        if (publicPrivateMoPair == null || publicPrivateMoPair.getRight() == null) {
-            ApiLogger.errorContext(SERVICE, NO_PAIR_FOUND, null);
+        if (publicPrivateMoPair == null) {
+            ApiLogger.errorContext(SERVICE, NO_PUBLIC_AND_NO_PRIVATE_MO_DATA_FOUND, null, logMap);
             return null;
         }
+        else {
+            if (publicPrivateMoPair.getLeft() == null) {
+                ApiLogger.errorContext(SERVICE, NO_PUBLIC_MO_DATA_FOUND, null, logMap);
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/ManagingOfficerChangeService.java:227]: continuing?

**Review by bmcgreevyCH — APPROVED** [review]

**Review by AndreKramer7654 — APPROVED** [review]

**Pull request merged**